autonomi/src/networking/driver/task_handler.rs

```diff
@@ -0,0 +1,347 @@
+
+    pub fn insert_query(&mut self, id: OutboundRequestId, task: NetworkTask) {
+        info!("New query: with OutboundRequestId({id}): {task:?}");
+        if let NetworkTask::GetQuote {
+            resp,
+            data_type,
+            peer,
+            ..
+        } = task
+        {
+            self.get_cost.insert(id, (resp, data_type, peer));
+        }
+    }
+
+    pub fn update_closest_peers(
+        &mut self,
+        id: QueryId,
+        res: Result<kad::GetClosestPeersOk, kad::GetClosestPeersError>,
+    ) -> Result<(), TaskHandlerError> {
+        let responder = self
+            .closest_peers
+            .remove(&id)
+            .ok_or(TaskHandlerError::UnknownQuery(format!("QueryId {id:?}")))?;
+
+        match res {
+            Ok(kad::GetClosestPeersOk { peers, .. }) => {
+                responder
+                    .send(Ok(peers))
+                    .map_err(|_| TaskHandlerError::NetworkClientDropped)?;
+            }
+            Err(kad::GetClosestPeersError::Timeout { key, peers }) => {
+                trace!(
+                    "QueryId({id}): GetClosestPeersError::Timeout {:?}, peers: {:?}",
+                    hex::encode(key),
+                    peers
+                );
+                responder
+                    .send(Err(NetworkError::GetClosestPeersTimeout))
+                    .map_err(|_| TaskHandlerError::NetworkClientDropped)?;
+            }
+        }
+        Ok(())
+    }
+
+    pub fn update_get_record(
+        &mut self,
+        id: QueryId,
+        res: Result<kad::GetRecordOk, kad::GetRecordError>,
+    ) -> Result<(), TaskHandlerError> {
+        match res {
+            Ok(kad::GetRecordOk::FoundRecord(record)) => {
+                trace!(
+                    "QueryId({id}): GetRecordOk::FoundRecord {:?}",
+                    PrettyPrintRecordKey::from(&record.record.key)
+                );
+                let holders = self.get_record_accumulator.entry(id).or_default();
+                if let Some(peer_id) = record.peer {
+                    holders.insert(peer_id, record.record);
+                }
+            }
+            Ok(kad::GetRecordOk::FinishedWithNoAdditionalRecord { .. }) => {
+                trace!("QueryId({id}): GetRecordOk::FinishedWithNoAdditionalRecord");
+                let (responder, holders) = self.take_responder_and_holders_for_task(id)?;
+                let peers = holders.keys().cloned().collect();
+                let records_uniq = holders.values().cloned().fold(Vec::new(), |mut acc, x| {
+                    if !acc.contains(&x) {
+                        acc.push(x);
+                    }
+                    acc
+                });
+
+                let res = match &records_uniq[..] {
+                    [] => responder.send(Ok((None, peers))),
+                    [one] => responder.send(Ok((Some(one.clone()), peers))),
+                    [_one, _two, ..] => responder.send(Err(NetworkError::SplitRecord(holders))),
```

Comment:
For mutable datatypes, situation of `split` could happen, especially with large scaled network with the mutable data got certain rounds of updates.

In such case, we shall `aggregate` the multiple versions.

Hence, here, we shall at least report back all versions

Reply:
The `SplitRecord` error is managed by the datatypes layer, and is different for each type.

Reply:
`NetworkError::SplitRecord(holders)` contains all versions and holders per version. The data layer can then use this data and do as it sees fit.